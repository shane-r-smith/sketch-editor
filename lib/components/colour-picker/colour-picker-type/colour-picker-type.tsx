import type { BaseColourPickerProps } from "../colour-picker.types";

import FormControl from "@mui/material/FormControl";
import Select, { type SelectChangeEvent } from "@mui/material/Select";
import MenuItem from "@mui/material/MenuItem";
import type { ColourType } from "../../../domain";

export function ColourPickerType({
  colour,
  type,
  onChange,
  ...rest
}: BaseColourPickerProps) {
  const handleChange = (event: SelectChangeEvent) => {
    onChange(colour, event.target.value as ColourType);
  };

  return (
    <FormControl>
      <Select size="small" value={type} onChange={handleChange} {...rest}>
        <MenuItem value={"HEX"}>HEX</MenuItem>
        <MenuItem value={"RGB"}>RGB</MenuItem>
      </Select>
    </FormControl>
  );
}
