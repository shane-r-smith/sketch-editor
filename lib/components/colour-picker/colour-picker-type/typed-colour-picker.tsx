import { HexColorPicker, RgbStringColorPicker } from "react-colorful";
import type { BaseColourPickerProps } from "../colour-picker.types";
import { useCallback } from "react";
import { colord } from "colord";

import styles from "./typed-colour-picker.styles.module.scss";

export function TypedColourPicker({
  type,
  colour,
  onChange,
  ...rest
}: BaseColourPickerProps) {
  const handleChange = useCallback(
    (colour: string) => {
      onChange(colour, type);
    },
    [onChange, type]
  );

  const _className = styles.editorColourPicker;

  if (type === "RGB") {
    return (
      <RgbStringColorPicker
        className={_className}
        onChange={handleChange}
        color={colord(colour).toRgbString()}
        {...rest}
      />
    );
  }

  return (
    <HexColorPicker
      className={_className}
      color={colord(colour).toHex()}
      onChange={handleChange}
      {...rest}
    />
  );
}
