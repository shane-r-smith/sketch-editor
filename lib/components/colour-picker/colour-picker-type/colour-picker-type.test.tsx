import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import { ColourPickerType } from "./colour-picker-type";
import type { ColourType } from "../colour-picker.types";

describe("<ColourPickerType />", () => {
  const defaultProps = {
    colour: "#ff0000",
    type: "HEX" as ColourType,
    onChange: vi.fn(),
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<ColourPickerType {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("shows HEX and RGB options", async () => {
    // Arrange
    const user = userEvent.setup();

    // Act
    const { getByRole, getAllByRole } = render(
      <ColourPickerType {...defaultProps} />
    );

    const select = getByRole("combobox");
    await user.click(select);

    // Assert
    const options = getAllByRole("option");
    expect(options.map((opt) => opt.textContent)).toEqual(["HEX", "RGB"]);
  });

  it("calls onChange with correct arguments when selecting RGB", async () => {
    // Arrange
    const user = userEvent.setup();
    const onChange = vi.fn();

    // Act
    const { getByRole, getByText } = render(
      <ColourPickerType {...defaultProps} onChange={onChange} />
    );

    const select = getByRole("combobox");
    await user.click(select);
    await user.click(getByText("RGB"));

    // Assert
    expect(onChange).toHaveBeenCalledWith(defaultProps.colour, "RGB");
  });

  it("calls onChange with correct arguments when selecting HEX", async () => {
    // Arrange
    const user = userEvent.setup();
    const onChange = vi.fn();

    // Act
    const { getByRole, getByText } = render(
      <ColourPickerType {...defaultProps} type="RGB" onChange={onChange} />
    );

    const select = getByRole("combobox");
    await user.click(select);
    await user.click(getByText("HEX"));

    // Assert
    expect(onChange).toHaveBeenCalledWith(defaultProps.colour, "HEX");
  });
});
