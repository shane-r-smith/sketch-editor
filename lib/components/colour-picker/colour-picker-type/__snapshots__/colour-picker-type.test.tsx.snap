// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ColourPickerType /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-sc8y68-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-trm6af-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="combobox"
        tabindex="0"
      >
        HEX
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        tabindex="-1"
        value="HEX"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1nf2c5d-MuiNotchedOutlined-root"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
